<a href="https://colab.research.google.com/github/SilkSherstka/hse_machine_learning/blob/main/%D0%90%D0%BD%D0%B4%D1%80%D0%BE%D0%BD%D0%BE%D0%B2%D0%B0_trees_hw.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Домашнее задание: cравнение древовидных алгоритмов на NLP-данных

**Задача:** oбучить разные древовидные модели на текстах новостей, замерить test accuracy и время обучения, заполнить таблицы.

## 1. Загрузка и подготовка данных (код дан)

In [1]:
from sklearn.datasets import fetch_20newsgroups
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
import time
import pandas as pd

# Загружаем 3 категории
categories = ['rec.sport.baseball', 'sci.space', 'comp.graphics']
newsgroups = fetch_20newsgroups(subset='all', categories=categories, shuffle=True, random_state=42)

# TF-IDF векторизация
vectorizer = TfidfVectorizer(max_features=5000, ngram_range=(1, 2), stop_words='english')
X = vectorizer.fit_transform(newsgroups.data)
y = newsgroups.target

# Разделение на train/test
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print(f"Размер обучающей выборки: {X_train.shape}")
print(f"Размер тестовой выборки: {X_test.shape}")

Размер обучающей выборки: (2363, 5000)
Размер тестовой выборки: (591, 5000)


## 2. Образец: Дерево решений (Decision Tree)

**Гиперпараметр:** `max_depth`

In [2]:
from sklearn.tree import DecisionTreeClassifier

results_tree = []

# Образец: мы делаем перебор значений max_depth = [3, 5, 10, 20, None])
# Для каждого: замер времени, обучение, accuracy на тесте, сохранение в results_tree

for depth in [3, 5, 10, 20, None]:
    start = time.time()
    clf = DecisionTreeClassifier(max_depth=depth, random_state=42)
    clf.fit(X_train, y_train)
    fit_time = time.time() - start
    acc = accuracy_score(y_test, clf.predict(X_test))
    results_tree.append({'max_depth': depth, 'test_accuracy': round(acc, 4), 'time_sec': round(fit_time, 2)})
    print(f"depth={depth}: acc={acc:.4f}, time={fit_time:.2f}s")

df_tree = pd.DataFrame(results_tree)
print("\nТаблица 1. Дерево решений")
print(df_tree.to_markdown(index=False))

depth=3: acc=0.6244, time=0.08s
depth=5: acc=0.6971, time=0.13s
depth=10: acc=0.7716, time=0.24s
depth=20: acc=0.8409, time=0.38s
depth=None: acc=0.8697, time=0.66s

Таблица 1. Дерево решений
|   max_depth |   test_accuracy |   time_sec |
|------------:|----------------:|-----------:|
|           3 |          0.6244 |       0.08 |
|           5 |          0.6971 |       0.13 |
|          10 |          0.7716 |       0.24 |
|          20 |          0.8409 |       0.38 |
|         nan |          0.8697 |       0.66 |


## 3. Заполните таблицы для остальных алгоритмов

По аналогии с образцом обучите следующие модели и запишите результаты.

### 3.1. Случайный лес (Random Forest)

**Гиперпараметр:** `max_depth` (фиксируем `n_estimators=100`)

| max_depth | test_accuracy | time_sec |
|-----------|---------------|----------|
| 5         |               |          |
| 10        |               |          |
| 20        |               |          |
| None      |               |          |

In [3]:
from sklearn.ensemble import RandomForestClassifier

results_rf = []

# Ваш код здесь (перебор max_depth)
for depth in [5, 10, 20, None]:
    start = time.time()
    clf = RandomForestClassifier(n_estimators=100, max_depth=depth, random_state=42)
    clf.fit(X_train, y_train)
    fit_time = time.time() - start
    acc = accuracy_score(y_test, clf.predict(X_test))
    results_rf.append({'max_depth': depth, 'test_accuracy': round(acc, 4), 'time_sec': round(fit_time, 2)})
    print(f"depth={depth}: acc={acc:.4f}, time={fit_time:.2f}s")


df_rf = pd.DataFrame(results_rf)
print("\nТаблица 2. Случайный лес")
print(df_rf.to_markdown(index=False))

depth=5: acc=0.9002, time=0.69s
depth=10: acc=0.9154, time=1.00s
depth=20: acc=0.9357, time=1.41s
depth=None: acc=0.9526, time=1.61s

Таблица 2. Случайный лес
|   max_depth |   test_accuracy |   time_sec |
|------------:|----------------:|-----------:|
|           5 |          0.9002 |       0.69 |
|          10 |          0.9154 |       1    |
|          20 |          0.9357 |       1.41 |
|         nan |          0.9526 |       1.61 |


### 3.2. Градиентный бустинг (Gradient Boosting)

**Гиперпараметр:** `learning_rate` (фиксируем `n_estimators=100, max_depth=3`)

| learning_rate | test_accuracy | time_sec |
|---------------|---------------|----------|
| 0.01          |               |          |
| 0.05          |               |          |
| 0.1           |               |          |
| 0.5           |               |          |

In [5]:
from sklearn.ensemble import GradientBoostingClassifier

results_gb = []

# Ваш код здесь (перебор learning_rate)
for lr in [0.01, 0.05, 0.1, 0.5]:
    start = time.time()
    clf = GradientBoostingClassifier(n_estimators=100, max_depth=3, learning_rate=lr, random_state=42)
    clf.fit(X_train, y_train)
    fit_time = time.time() - start
    acc = accuracy_score(y_test, clf.predict(X_test))
    results_gb.append({'learning_rate': lr, 'test_accuracy': round(acc, 4), 'time_sec': round(fit_time, 2)})
    print(f"learning_rate={lr}: acc={acc:.4f}, time={fit_time:.2f}s")


df_gb = pd.DataFrame(results_gb)
print("\nТаблица 3. Градиентный бустинг")
print(df_gb.to_markdown(index=False))

learning_rate=0.01: acc=0.8782, time=22.90s
learning_rate=0.05: acc=0.9475, time=22.48s
learning_rate=0.1: acc=0.9577, time=23.10s
learning_rate=0.5: acc=0.9662, time=21.78s

Таблица 3. Градиентный бустинг
|   learning_rate |   test_accuracy |   time_sec |
|----------------:|----------------:|-----------:|
|            0.01 |          0.8782 |      22.9  |
|            0.05 |          0.9475 |      22.48 |
|            0.1  |          0.9577 |      23.1  |
|            0.5  |          0.9662 |      21.78 |


### 3.3. XGBoost

**Гиперпараметр:** `gamma` (фиксируем `n_estimators=100, learning_rate=0.1`)

| gamma | test_accuracy | time_sec |
|-------|---------------|----------|
| 0     |               |          |
| 0.1   |               |          |
| 0.5   |               |          |
| 1.0   |               |          |

In [6]:
from xgboost import XGBClassifier

results_xgb = []

# Ваш код здесь (перебор gamma, use_label_encoder=False, eval_metric='logloss')
for gamma in [0, 0.1, 0.5, 1.0]:
    start = time.time()
    clf = XGBClassifier(n_estimators=100, learning_rate=0.1, gamma=gamma, random_state=42, use_label_encoder=False, eval_metric='logloss')
    clf.fit(X_train, y_train)
    fit_time = time.time() - start
    acc = accuracy_score(y_test, clf.predict(X_test))
    results_xgb.append({'gamma': gamma, 'test_accuracy': round(acc, 4),'time_sec': round(fit_time, 2)})
    print(f"gamma={gamma}: acc={acc:.4f}, time={fit_time:.2f}s")


df_xgb = pd.DataFrame(results_xgb)
print("\nТаблица 4. XGBoost")
print(df_xgb.to_markdown(index=False))

/usr/local/lib/python3.12/dist-packages/xgboost/training.py:200: UserWarning: [20:30:43] WARNING: /__w/xgboost/xgboost/src/learner.cc:782: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)


gamma=0: acc=0.9459, time=24.02s


/usr/local/lib/python3.12/dist-packages/xgboost/training.py:200: UserWarning: [20:31:07] WARNING: /__w/xgboost/xgboost/src/learner.cc:782: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)


gamma=0.1: acc=0.9442, time=23.43s


/usr/local/lib/python3.12/dist-packages/xgboost/training.py:200: UserWarning: [20:31:31] WARNING: /__w/xgboost/xgboost/src/learner.cc:782: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)


gamma=0.5: acc=0.9492, time=22.85s


/usr/local/lib/python3.12/dist-packages/xgboost/training.py:200: UserWarning: [20:31:54] WARNING: /__w/xgboost/xgboost/src/learner.cc:782: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)


gamma=1.0: acc=0.9492, time=21.07s

Таблица 4. XGBoost
|   gamma |   test_accuracy |   time_sec |
|--------:|----------------:|-----------:|
|     0   |          0.9459 |      24.02 |
|     0.1 |          0.9442 |      23.43 |
|     0.5 |          0.9492 |      22.85 |
|     1   |          0.9492 |      21.07 |


### 3.4. AdaBoost

**Гиперпараметр:** `learning_rate` (фиксируем `n_estimators=100`)

| learning_rate | test_accuracy | time_sec |
|---------------|---------------|----------|
| 0.5           |               |          |
| 1.0           |               |          |
| 1.5           |               |          |
| 2.0           |               |          |

In [7]:
from sklearn.ensemble import AdaBoostClassifier

results_ada = []

# Ваш код здесь (перебор learning_rate)
for lr in [0.5, 1.0, 1.5, 2.0]:
    start = time.time()
    clf = AdaBoostClassifier(n_estimators=100, learning_rate=lr, random_state=42)
    clf.fit(X_train, y_train)
    fit_time = time.time() - start
    acc = accuracy_score(y_test, clf.predict(X_test))
    results_ada.append({'learning_rate': lr, 'test_accuracy': round(acc, 4), 'time_sec': round(fit_time, 2)})
    print(f"learning_rate={lr}: acc={acc:.4f}, time={fit_time:.2f}s")

df_ada = pd.DataFrame(results_ada)
print("\nТаблица 5. AdaBoost")
print(df_ada.to_markdown(index=False))

learning_rate=0.5: acc=0.8629, time=4.99s
learning_rate=1.0: acc=0.9205, time=4.32s
learning_rate=1.5: acc=0.9255, time=3.90s
learning_rate=2.0: acc=0.8731, time=4.27s

Таблица 5. AdaBoost
|   learning_rate |   test_accuracy |   time_sec |
|----------------:|----------------:|-----------:|
|             0.5 |          0.8629 |       4.99 |
|             1   |          0.9205 |       4.32 |
|             1.5 |          0.9255 |       3.9  |
|             2   |          0.8731 |       4.27 |


### 3.5. LightGBM

**Гиперпараметр:** `num_leaves` (фиксируем `n_estimators=100, learning_rate=0.1`)

| num_leaves | test_accuracy | time_sec |
|------------|---------------|----------|
| 15         |               |          |
| 31         |               |          |
| 63         |               |          |
| 127        |               |          |

In [8]:
from lightgbm import LGBMClassifier

results_lgbm = []

# Ваш код здесь (перебор num_leaves, n_jobs=-1)
for num_leaves in [15, 31, 63, 127]:
    start = time.time()
    clf = LGBMClassifier(n_estimators=100, learning_rate=0.1, num_leaves=num_leaves, random_state=42, n_jobs=-1)
    clf.fit(X_train, y_train)
    fit_time = time.time() - start
    acc = accuracy_score(y_test, clf.predict(X_test))
    results_lgbm.append({'num_leaves': num_leaves, 'test_accuracy': round(acc, 4), 'time_sec': round(fit_time, 2)})
    print(f"num_leaves={num_leaves}: acc={acc:.4f}, time={fit_time:.2f}s")


df_lgbm = pd.DataFrame(results_lgbm)
print("\nТаблица 6. LightGBM")
print(df_lgbm.to_markdown(index=False))

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.171548 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 61063
[LightGBM] [Info] Number of data points in the train set: 2363, number of used features: 2963
[LightGBM] [Info] Start training from score -1.119999
[LightGBM] [Info] Start training from score -1.093126
[LightGBM] [Info] Start training from score -1.083076


/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(


num_leaves=15: acc=0.9662, time=5.88s
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.067528 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 61063
[LightGBM] [Info] Number of data points in the train set: 2363, number of used features: 2963
[LightGBM] [Info] Start training from score -1.119999
[LightGBM] [Info] Start training from score -1.093126
[LightGBM] [Info] Start training from score -1.083076


/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(


num_leaves=31: acc=0.9594, time=5.38s
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.064700 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 61063
[LightGBM] [Info] Number of data points in the train set: 2363, number of used features: 2963
[LightGBM] [Info] Start training from score -1.119999
[LightGBM] [Info] Start training from score -1.093126
[LightGBM] [Info] Start training from score -1.083076
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf


/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(


num_leaves=63: acc=0.9577, time=10.45s
[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.074793 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 61063
[LightGBM] [Info] Number of data points in the train set: 2363, number of used features: 2963
[LightGBM] [Info] Start training from score -1.119999
[LightGBM] [Info] Start training from score -1.093126
[LightGBM] [Info] Start training from score -1.083076
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain,

/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(


### 3.6. CatBoost

**Гиперпараметр:** `depth` (фиксируем `iterations=100, learning_rate=0.1`)

| depth | test_accuracy | time_sec |
|-------|---------------|----------|
| 3     |               |          |
| 5     |               |          |
| 7     |               |          |
| 10    |               |          |

In [11]:
!pip install catboost

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 97.1/97.1 MB 7.7 MB/s eta 0:00:00


In [12]:
from catboost import CatBoostClassifier

results_cat = []

# Ваш код здесь (перебор depth, verbose=0)
for depth in [3, 5, 7, 10]:
    start = time.time()
    clf = CatBoostClassifier(iterations=100, learning_rate=0.1, depth=depth, random_state=42, verbose=0)
    clf.fit(X_train, y_train)
    fit_time = time.time() - start
    acc = accuracy_score(y_test, clf.predict(X_test))
    results_cat.append({'depth': depth, 'test_accuracy': round(acc, 4), 'time_sec': round(fit_time, 2)})
    print(f"depth={depth}: acc={acc:.4f}, time={fit_time:.2f}s")

df_cat = pd.DataFrame(results_cat)
print("\nТаблица 7. CatBoost")
print(df_cat.to_markdown(index=False))

depth=3: acc=0.9340, time=9.45s
depth=5: acc=0.9425, time=23.91s
depth=7: acc=0.9442, time=73.65s
depth=10: acc=0.9526, time=535.72s

Таблица 7. CatBoost
|   depth |   test_accuracy |   time_sec |
|--------:|----------------:|-----------:|
|       3 |          0.934  |       9.45 |
|       5 |          0.9425 |      23.91 |
|       7 |          0.9442 |      73.65 |
|      10 |          0.9526 |     535.72 |


## 4. Итоговая таблица лучших результатов

In [13]:
# Соберите лучшие результаты из каждой таблицы вручную или автоматически
# Пример структуры:

summary = pd.DataFrame([
    ('Decision Tree', 'max_depth=None', 0.8697, 0.66),
    ('Random Forest', 'max_depth=None', 0.9526, 1.61),
    ('Gradient Boosting', 'learning_rate=0.5', 0.9662, 21.78),
    ('XGBoost', 'gamma=0.1', 0.9492, 21.07),
    ('AdaBoost', 'learning_rate=1.5', 0.9255, 3.90),
    ('LightGBM', 'num_leaves=15', 0.9662, 5.88),
    ('CatBoost', 'depth=10', 0.9526, 535.72),
], columns=['Algorithm', 'Best params', 'Best test acc', 'Time (sec)'])

print("Итоговая таблица лучших результатов")
print(summary.to_markdown(index=False))

Итоговая таблица лучших результатов
| Algorithm         | Best params       |   Best test acc |   Time (sec) |
|:------------------|:------------------|----------------:|-------------:|
| Decision Tree     | max_depth=None    |          0.8697 |         0.66 |
| Random Forest     | max_depth=None    |          0.9526 |         1.61 |
| Gradient Boosting | learning_rate=0.5 |          0.9662 |        21.78 |
| XGBoost           | gamma=0.5         |          0.9492 |        22.85 |
| AdaBoost          | learning_rate=1.5 |          0.9255 |         3.9  |
| LightGBM          | num_leaves=15     |          0.9662 |         5.88 |
| CatBoost          | depth=10          |          0.9526 |       535.72 |


## 5. Вопросы (ответить текстом в следующей ячейке)

1. Какой алгоритм показал максимальную точность? Какие параметры к этому привели?

2. Какой алгоритм быстрее всего обучался? Во сколько раз он быстрее самого медленного?

3. У каких алгоритмов наблюдалось переобучение? При каких параметрах?

4. Какой алгоритм вы выбрали бы для продакшн-системы с миллионом текстов? Почему?

# Напишите свои ответы здесь

**Ответ 1:**
Gradient Boosting (learning_rate = 0.5) и LightGBM (num_leaves=15), точность = 0.9662

**Ответ 2:**
Быстрее всего обучался Decision Tree - при лучших параметрах (max_depth=None) за 0.66 сек.
Самый медленный - CatBoost (depth=10) - 535.72 сек. При лучших параметрах Decision Tree быстрее CatBoost примерно в 811 раз.

**Ответ 3:**
Adaboost заметно переобучился - при learning_rate = 2.0 точность упала до 0.8731 (при лучшей точности = 0.9255)

Немного переобучился lightgbm (точность упала с 0.9662 до 0.9577)


**Ответ 4:**
LightGBM - лучшая точность при довольно быстром обучении (плюс не попал в топ по переобучению)

## Критерии оценки

| Что оценивается | Баллы |
|----------------|-------|
| Заполнены все 7 таблиц (правильно собраны accuracy и время) | 3 |
| Заполнена итоговая таблица лучших результатов | 1 |
| Ответы на 4 вопроса (по 0.5 балла) | 2 |
| Код воспроизводим (random_state=42, порядок ячеек корректен) | 1 |
| **Качество кода** (отсутствие дублирования, осмысленные имена переменных, циклы вместо копипасты, использование .to_markdown()) | 3 |
| **Итого** | **10** |

### Детали по качеству кода (+3 балла):
- **+1** — использование единого шаблона для всех экспериментов (цикл + list.append + pd.DataFrame)
- **+1** — правильное форматирование вывода (округление времени и accuracy до 2-4 знаков)
- **+1** — читаемые имена переменных, комментарии, отсутствие `eval()`, `exec()` и магических чисел

**Штрафы:**
- -1 балл, если код не запускается без ошибок
- -1 балл, если отсутствует `random_state=42` в любой из моделей
- -1 балл, если таблицы выведены криво (не через `to_markdown` или нечитаемый `print`)